src/stan/model/indexing/rvalue.hpp

```diff
@@ -27,115 +28,186 @@ namespace model {
  * @return Input value.
  */
 template <typename T>
-inline T rvalue(const T& c, const nil_index_list& /*idx*/,
-                const char* /*name*/ = "", int /*depth*/ = 0) {
-  return c;
+inline T rvalue(T&& c, const nil_index_list& /*idx*/, const char* /*name*/ = "",
+                int /*depth*/ = 0) {
+  return std::forward<T>(c);
 }
 
 /**
- * Return the result of indexing the specified Eigen vector with a
- * sequence containing one single index, returning a scalar.
+ * Return the result of indexing a type without taking a subset. Mostly used as
+ * an intermediary rvalue function when doing multiple subsets.
  *
- * Types:  vec[single] : scal
+ * Types:  type[omni] : type
  *
- * @tparam T Scalar type.
- * @param[in] v Vector being indexed.
- * @param[in] idx One single index.
+ * @tparam T Any type.
+ * @param[in] an object.
+ * @param[in] idx Index consisting of one omni-index.
  * @param[in] name String form of expression being evaluated.
  * @param[in] depth Depth of indexing dimension.
- * @return Result of indexing vector.
+ * @return Result of indexing matrix.
  */
 template <typename T>
-inline T rvalue(const Eigen::Matrix<T, Eigen::Dynamic, 1>& v,
-                const cons_index_list<index_uni, nil_index_list>& idx,
+inline T rvalue(T&& a, const cons_index_list<index_omni, nil_index_list>& idx,
```

Comment:
One thing I want to call out here, I actually don't want to have to do the Type&& perfect forwarding syntax here, but if I do it here and then have const & everywhere else there's ambiguity errors with the overloads and the ones that have a template parameter in the `cons_index_list<>` like `cons_index_list<L, nil_index_list>`